src/diagnostic/GraphProbe.cpp

```diff
@@ -0,0 +1,420 @@
+/** Copyright 2017-2021 CNRS-AIST JRL and CNRS-UM LIRMM */
+
+#include <tvm/diagnostic/GraphProbe.h>
+
+#include <iomanip>
+#include <iostream>
+#include <queue>
+#include <string>
+
+namespace
+{
+using namespace tvm::diagnostic;
+using namespace tvm::graph::internal;
+
+std::string indent(int val) { return std::string(val, ' '); }
+
+// shared code for any and none (see below)
+template<bool b>
+bool check(const std::vector<GraphProbe::OutputVal> & outputVal, std::function<bool(const Eigen::MatrixXd &)> select)
+{
+  for(const auto & v : outputVal)
+  {
+    if(select(std::get<Eigen::MatrixXd>(v)))
+      return b;
+  }
+  return !b;
+}
+
+// any(outputVal, select) returns true if any value v in outputVal is such that select(v) is true, and false otherwise
+constexpr auto any = check<true>;
+// none(outputVal, select) returns true if no value v in outputVal is such that select(v) is true, and false otherwise
+constexpr auto none = check<false>;
+
+Eigen::MatrixXd transposeIfVector(const Eigen::MatrixXd & M)
+{
+  if(M.cols() == 1)
+    return M.transpose();
+  else
+    return M;
+}
```

Comment:
Given that is only used for outputs and to avoid unnecessary copies:

```c++
auto transposeIfVector(const Eigen::MatrixXd & M, const Eigen::IOFormat & fmt = Eigen::IOFormat{})
{
  if(M.cols() == 1)
  {
    return M.transpose().format(fmt);
  }
  return M.format(fmt);
}
```

Reply:
This was not possible this way as the return types differ. But I did a better version with two different ` Eigen::Format`